### PR TITLE
Add tensor forest classification train in core

### PR DESCRIPTION
**Pull request by yupbank**
According to https://github.com/tensorflow/community/blob/master/rfcs/20180626-tensor-forest.md

This is a first step. And the whole process is tracked by #21830

We are adding an canned TensorForestClassifier and TensorForestRegressor into core.

The Pr is only a third step implementing TensorForestClassifier with train only functionality.

### CODE REVIEW

File: tensorflow/core/kernels/tensor_forest/resource_ops.cc [hunk 1]
```diff
@@ -114,22 +115,115 @@ class TensorForestTreeSizeOp : public OpKernel {
   }
 };
 
+class TensorForestCreateFertileStatsVariableOp : public OpKernel {
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resource_ops.cc]: please add a comment to as what this represents. Eg a resource around statistics of the leaves for all the leaves in a tree that can still be splitted

File: tensorflow/core/kernels/tensor_forest/resource_ops.cc [hunk 2]
```diff
@@ -114,22 +115,115 @@ class TensorForestTreeSizeOp : public OpKernel {
...
+    // Only create one, if one does not exist already. Report status for all
+    // other exceptions.
+    auto status = CreateResource(context, HandleFromInput(context, 0), result);
+    if (!status.ok() && status.code() != tensorflow::error::ALREADY_EXISTS) {
+      OP_REQUIRES(context, false, status);
+    }
+  }
+};
+
+// Op for serializing a model.
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resource_ops.cc]: probably a copy past comment. for serializing fertile leaf stats

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 3]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+void TensorForestTreeResource::SplitNode(const int32 node,
+                                         boosted_trees::Vector& left_leaf,
+                                         boosted_trees::Vector& right_leaf,
+                                         std::vector<int32>* new_children) {
+  using boosted_trees::Leaf;
+  using boosted_trees::Node;
+  auto current_split =
+      decision_tree_->mutable_nodes(node)->mutable_dense_split();
+  current_split->Swap(best->mutable_split());
+  // add left leaf
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: where is best coming from??

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 4]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+  auto* new_left = decision_tree_->add_nodes();
+  new_left->mutable_leaf()->mutable_vector()->Swap(left_leaf);
+  new_children->push_back(decision_tree_->nodes_size());
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: nodes_size - 1?

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 5]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+  // add right leaf
+  auto* new_right = decision_tree_->add_nodes();
+  new_right->mutable_leaf()->mutable_vector()->Swap(right_leaf);
+  new_children->push_back(decision_tree_->nodes_size());
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: ditto

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 6]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+  if (fertile_stats_->node_to_slot().count(node_id) > 0) {
+    auto slot = fertile_stats_->node_to_slot().at(node_id);
+    return slot.leaf_stats().weight_sum() > 0 &&
+           slot.candidates_size() > splits_to_consider;
+  } else {
+    return false;
+  }
+};
+
+const bool TensorForestFertileStatsResource::IsSlotFinished(
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: can u add comments to these methods to help understand what they are for

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 7]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+    const int32 splits_to_consider) const {
+  if (IsSlotInitialized(node_id, splits_to_consider)) {
+    auto slot = fertile_stats_->node_to_slot().at(node_id);
+    return slot.leaf_stats().weight_sum() > split_nodes_after_samples;
+  }
+  return false;
+};
+
+void TensorForestFertileStatsResource::UpdateSlotStats(
+    const bool is_regression, const int32 node_id, const int32 example_id,
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: lets stick to node_id everywhere? in a previous file u call them node but it is still an int

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 8]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+  }
+};
+
+const bool TensorForestFertileStatsResource::BestSplitFromSlot(
+    const tensor_forest::FertileSlot& slot,
+    tensor_forest::SplitCandidate* best) const {
+  float min_score = FLT_MAX;
+  for (int i = 0; i < slot.candidates_size(); i++) {
+    auto candidate = slot->candidates(i);
+    // for (auto candidate : slot.candidates()) {
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: const auto&?

Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: is the comment on line 195 important

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 9]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+    float left_gini =
+        1 - candidate.left_split_stats().sum_of_square().value(0) /
+                (candidate.left_split_stats().sum().value(0) *
+                 candidate.left_split_stats().sum().value(0));
+    float right_gini =
+        1 - candidate.right_split_stats().sum_of_square().value(0) /
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: can u make getGini a method that takes in split stats? This way it is much shorter and much less error prone

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 10]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+          0, candidate.right_split_stats().sum_of_square().value(0) -
+                 old_right_weight * old_right_weight + new_weight * new_weight);
+    }
+  }
+};
+
+const bool TensorForestFertileStatsResource::BestSplitFromSlot(
+    const tensor_forest::FertileSlot& slot,
+    tensor_forest::SplitCandidate* best) const {
+  float min_score = FLT_MAX;
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: comment: we are looking to minimize the gini

File: tensorflow/core/kernels/tensor_forest/resources.cc [hunk 11]
```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
...
+    const int32 example_id, const int32 num_targets,
+    const TTypes<float>::ConstMatrix* dense_feature,
+    const TTypes<float>::ConstMatrix* labels) {
+  auto slot = fertile_stats_->node_to_slot().at(node_id);
+  slot.mutable_leaf_stats()->set_weight_sum(slot.leaf_stats().weight_sum() + 1);
+  auto candidate = slot.add_candidates();
+  auto split = candidate->mutable_split();
+  split->set_threshold(threshold);
+  split->set_feature_id(feature_id);
+  float incoming_weight = 1.0;
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/resources.cc]: it looks like it is just a cont, move it to the top of the file, make it a real const

**Review by nataliaponomareva — COMMENTED** [review]
I wonder whether it is still WIP. For one, i dont see the graph so it is hard to comment overall

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @nataliaponomareva: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @nataliaponomareva: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @nataliaponomareva: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/core/kernels/tensor_forest/tensor_forest.proto [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+option java_outer_classname = "TensorForestProtos";
+option java_multiple_files = true;
+option java_package = "org.tensorflow.framework";
+
+import "tensorflow/core/kernels/boosted_trees/boosted_trees.proto";
+
+message SplitStats {
+  // For classification,
+  // we store the sum of example weights and the sum of the squares of
+  // example weights.  Gini(X) = 1 - \sum_i{(X_i/n)^2}
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/tensor_forest.proto]: sorry i don't follow completely. X_i is the mean here? What is i index then?

File: tensorflow/core/kernels/tensor_forest/tensor_forest.proto [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+  // Right gini stats.
+  SplitStats right_split_stats = 3;
+
+  // Right counts are inferred from FertileSlot.post_init_leaf_stats and
+  // left_stats.
+  LeafStat left_leaf_stats = 4;
+
+  // Right stats (not full counts) are kept here.
+  // LeafStat right_stats = 5;
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/tensor_forest.proto]: commented out?

File: tensorflow/core/kernels/tensor_forest/train_ops.cc [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+      return;
+    }
+
+    auto* worker_threads = context->device()->tensorflow_cpu_worker_threads();
+    const int32 num_threads = worker_threads->num_threads;
+    const int64 cost_per_traverse = 500;
+    auto traverse = [&out, &data_set, decision_tree_resource](int64 start,
+                                                              int64 end) {
+      for (int example_id = start; example_id < end; ++example_id) {
+        out(example_id) =
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/train_ops.cc]: didn't we have this op already for the inference?

File: tensorflow/core/kernels/tensor_forest/train_ops.cc [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+    const Tensor* labels_t = nullptr;
+    OP_REQUIRES_OK(context, context->input("labels", &labels_t));
+    const Tensor* leaf_ids_t = nullptr;
+    OP_REQUIRES_OK(context, context->input("leaf_ids", &leaf_ids_t));
+
+    TensorForestFertileStatsResource* fertile_stats_resource;
+    OP_REQUIRES_OK(context, LookupResource(context, HandleFromInput(context, 0),
+                                           &fertile_stats_resource));
+
+    mutex_lock l(*fertile_stats_resource->get_mutex());
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/train_ops.cc]: can it be shared lock here

File: tensorflow/core/kernels/tensor_forest/train_ops.cc [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+
+    // Create one mutex per leaf. We need to protect access to leaf pointers,
+    // so instead of grouping examples by leaf, we spread examples out among
+    // threads to provide uniform work for each of them and protect access
+    // with mutexes.
+    std::unordered_map<int, std::unique_ptr<mutex>> locks;
+    for (int i = 0; i < batch_size; ++i) {
+      const int32 leaf_id = leaf_ids(i);
+      if (locks.find(leaf_id) == locks.end()) {
+        // TODO(gilberth): Consider using a memory pool for these.
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/train_ops.cc]: please remove todo for gilberth

File: tensorflow/core/kernels/tensor_forest/train_ops.cc [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+    const int32 num_targets = is_regression_ ? (std::max(1, label_dim)) : 1;
+
+    // Ids of leaves that can split.
+    std::unordered_set<int32> ready_to_split;
+    mutex set_lock;
+
+    auto worker_threads = context->device()->tensorflow_cpu_worker_threads();
+    int num_threads = worker_threads->num_threads;
+
+    // TODO(gilberth): This is a rough approximation based on measurements
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/train_ops.cc]: ditto here. i assume it means this code is a direct copy from contrib? :)

File: tensorflow/core/kernels/tensor_forest/train_ops.cc [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+    auto output = output_finished_t->unaligned_flat<int32>();
+    std::copy(ready_to_split.begin(), ready_to_split.end(), output.data());
+  };
+
+  void AddExample(const int32 num_targets, const int32 leaf_id,
+                  const int32 example_id,
+                  const TTypes<float>::ConstMatrix* dense_features,
+                  const TTypes<float>::ConstMatrix* labels,
+                  std::unordered_map<int, std::unique_ptr<mutex>>* locks,
+                  std::queue<std::tuple<int32, int32>>* waiting,
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/tensor_forest/train_ops.cc]: what is this waiting?

**Review by nataliaponomareva — CHANGES_REQUESTED** [review]
Thanks Peng, it looks mostly good!
My two main questions
1) SplitStats for classification - didn't understand why you need sum/sum of squares there
2) Tests for the ops/training?

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**